### PR TITLE
Build CUTEst inside deps/

### PR DESCRIPTION
**Pull request by timholy**
The first commit downloads the files into a directory `deps/files`. After checking for a "global" installation (ref https://github.com/JuliaSmoothOptimizers/CUTEst.jl/issues/103#issuecomment-262296220), it then checks for a "local" (julia-specific) install. Only if the latter is not found does it download and install the files. EDIT: this will avoid unnecessary re-installations whenever you tag a new version of CUTEst (`Pkg.build` runs every time a package gets updated).

The second commit may be more controversial. It adds a new environment variable, `CUTEst problems` which points to the directory `deps/files/problems`. It leverages this to build all the `*.so` files (and the `AUTOMAT.d` and `OUTSDIF.d` files) inside this folder, rather than the user's current folder. I've been deleting `.so` files from various folders that I run CUTEst problems in, so centralizing these files seems like a nice way to prevent clutter. On the other hand, I wonder if it's currently possible to have two different julia sessions running in two different directories, both working on CUTEst problems; if this gets centralized, that won't be possible anymore (I presume). If so, I guess one option would be to build them in a temporary folder, and have each julia session generate a different random temporary.

### CODE REVIEW

**Comment by timholy** [comment]
I should add, this fixes #103.

**Comment by timholy** [comment]
The OSX errors arise from installation (click to expand the `julia -E 'Pkg.build("CUTEst");'`), specifically a `Error: no implicit conversion of nil into String` apparently in the Homebrew ruby code. I don't think that's related to this PR, but perhaps someone more knowledgeable than me about OSX should take a peek?

**Comment by dpo** [comment]
Must be a recent change in Homebrew. I'll have a look asap.

**Comment by dpo** [comment]
I am wondering if the issue comes from Travis/OSX because I can't reproduce locally. I "fixed" the Homebrew formula for CUTEst to make Travis happy and now it installs, but there's an error further down, in the tests. I can't reproduce them locally either.

**Comment by abelsiqueira** [comment]
I restarted CUTEst develop branch and it works for 0.4 and 0.5. There is something here that is breaking OSX. Look at [this line](https://travis-ci.org/abelsiqueira/CUTEst.jl/jobs/183338424#L591). It shows that CUTEst on OSX was installed on a tmp dir.

File: deps/build.jl [hunk 1]
```diff
@@ -34,39 +35,58 @@ if validate_libcutest()
...
+          prefix = Homebrew.prefix(p)
+          open("$prefix/$p.bashrc") do f
+            for line in readlines(f)
+              var = split(split(line, "=")[1])[2]
+              path = chomp(split(line, "=")[2])
+              write(cenv, "ENV[\"$var\"] = \"$path\"\n")
+            end
           end
         end
+        println(cenv, "ENV[\"CUTEst problems\"] = \"$(pwd())/problems\"")
```
Comment (abelsiqueira) [inline deps/build.jl]: This pwd() is possibly wrong.

File: deps/build.jl [hunk 2]
```diff
@@ -34,39 +35,58 @@ if validate_libcutest()
...
       end
     end
-  end
 
-  @static if is_linux()
-    lnxurl = "https://raw.githubusercontent.com/abelsiqueira/linux-cutest/master/install.sh"
-    run(`wget $lnxurl -O install.sh`)
-    run(`bash install.sh --install-deps`)
+    @static if is_linux()
+      cd(here) do
```
Comment (abelsiqueira) [inline deps/build.jl]: This is missing for OSX

File: deps/build.jl [hunk 3]
```diff
@@ -34,39 +35,58 @@ if validate_libcutest()
...
+            open("cutest_env.bashrc") do f
+              for p in ["PATH", "MANPATH", "LD_LIBRARY_PATH"]
+                write(cenv, "$p = get(ENV, \"$p\", \"\")\n")
+              end
+              for line in readlines(f)
+                var = split(split(line, "=")[1])[2]
+                path = chomp(split(line, "=")[2])
+                write(cenv, "ENV[\"$var\"] = \"$path\"\n")
+              end
+              println(cenv, "ENV[\"CUTEst problems\"] = \"$(pwd())/problems\"")
```
Comment (dpo) [inline deps/build.jl]: I don't think this is necessary; all problems in that folder are already in `ENV["MASTSIF"]`.

Comment (abelsiqueira) [inline deps/build.jl]: This is for the .so, .f and .d files, I believe, so that the workdir remains clear.

**Comment by abelsiqueira** [comment]
I made a [debug commit](https://travis-ci.org/abelsiqueira/CUTEst.jl/jobs/183541450) and indeed the problem is what's mentioned above: The `cd(tmpdir)` alone and the wrong `pwd()`. @timholy, do you wanna fix these, or can I merge locally and fix? @dpo, do you agree? I would then push directly after.

**Comment by dpo** [comment]
Yes, that seems right. Thanks @abelsiqueira!

**Comment by abelsiqueira** [comment]
Pushed with fix. Thanks @timholy .

**Comment by timholy** [comment]
Thanks for implementing the fix! I got busy with other things.

**Pull request closed without merging**